posts/how-to-make-a-composer-package.md

```diff
@@ -0,0 +1,552 @@
+
+## Implement the package class
+
+Until now, we have only configured the package and written the tests. Now, we need to actual implement the package class. We will start by creating a new file in the `src/` folder and name it `SaidQuote.php`. After that, copy the following code to the beginning of the file:
+
+<PrismSyntaxHighlighter
+input={`<?php
+ 
+namespace Tombenevides\SaidQuotes;
+ 
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\ClientException;
+ 
+class SaidQuote
+{
+    const BASE_ENDPOINT = "https://api.quotable.io/random";
+    private Client $client;
+ 
+    public function __construct(
+        private ?string $author = null,
+        ?Client $client = null
+    )
+    {
+        $this->client = $client ?? new Client();
+    }
+ 
+    public function getQuote()
+    {
+        try{
+            if(empty($this->author)){
+                throw new \Exception("An author name is required");
+            }
+  
+            $response = $this->client->get(self::BASE_ENDPOINT, [
+                'query' => [
+                    'author' => str_replace(' ', '-', $this->author)
+                ]
+            ])->getBody()->getContents();
+ 
+            $quote = json_decode($response, true);
+ 
+            return $quote['content'].' by '.ucwords($this->author);
+        }catch(ClientException $e){
+            return 'Error: Could not find any matching quotes for '.$this->author;
+        }catch(\Exception $e){
+            return 'Error: '.$e->getMessage();
+        }	
+    }
+ 
+}`}
+language='php'
+/>
+
+Let's go through what this code achieves. First, we set the namespace using the same one we defined in the `composer.json` file. 
+
+Then, we created the class using `SaidQuote` as the class name. In the class constructor we will receive two parameters. The first one is the author name and the second one is the Guzzle client. The class will also have a constant for the base endpoint of the API. 
+
+After that, we have our `getQuote()` function. The `try`-`catch` inside the function will handle the exceptions to return an appropriate message. Inside the `try` block, we will check if the author name is empty. If it is, we will throw an exception. If it is not, we will make a request to the API using the Guzzle client. In this request, we pass the author name as a parameter using a snake-case format. Finally, we decode the response and return the content concatenated with the author name.
+
+Now, if we run the tests, we will confirm that the package works as expected. Return back to your project's root directory before running the following,
+
+<Highlighter
+input={`./sg vendor/bin/pest
+ 
+ PASS  Tests\SaidQuoteTest
+  ✓ it returns a Jane Austen quote
+  ✓ it returns a message when it doesnt find a matching quote.
+  ✓ it returns a message when was not given an author name
+ 
+  Tests:  3 passed
+  Time:   0.02s`}
+language='bash'
+/>
```

Comment:
I'm getting this output:

```

   Whoops\Exception\ErrorException 

  The use statement with non-compound name 'GuzzleHttpClient' has no effect

  at tests/QuoteTest.php:3
      1▕ <?php
      2▕  
  ➜   3▕ use GuzzleHttpClient;
      4▕ use GuzzleHttpHandlerMockHandler;
      5▕  
      6▕ beforeEach(function () {
      7▕     $this->mockHandler = new MockHandler();
      8▕  
      9▕     $this->mockedClient = new Client([
```

I noticed that your `use` statements are different in your `said-quotes` repo on GitHub. Do you know what the issue may be here?

Reply:
Yes... Now that you've pointed out, all imports and namespace are without the backslashes. I don't understand why, but the current snippet in this PR is backslashed, but when I look at the preview link, it's not. I noticed that I didn't import all the classes used in the tests, so I pushed a commit to fix that. However, the error you showed me is about something I don't think I can control unless the script that converts Markdown to html ignores the backslashes. Perhaps in this case, adding an extra backslash will solve the problem, but it doesn't make much sense as the error occurs within a codeblock section, so, I can add the extra backslashes if you confirm that this is the problem.

Reply:
![image](https://user-images.githubusercontent.com/18440704/143433208-bb9e2c74-cc52-48b0-8de1-5aa975d1c0ca.png)

I did a test and as you can see it works with an extra backslash. Unfortunately, it also adds a space before and after the slash. I've tried looking at the repo of the package you guys use to see if they provide an example for escaping that kind of special character, but I haven't found any examples. So if someone who knows more about the package can show us the best way to escape backslashes without adding extra spaces, that would be nice. Until then, I'll test the package later and see if I can find any information.

Reply:
We will figure this out on our end, thanks for updating this!

Reply:
@tombenevides — feel free to start on the next tutorial, we will get this published by resolving the markdown bug

Reply:
@katjuell I created [DEVED-278](https://sourcegraph.atlassian.net/browse/DEVED-278?atlOrigin=eyJpIjoiYmY3OWUzOGU2YmU0NDRkNzlmODhiODJmNDExZjM3MWMiLCJwIjoiaiJ9) with this bug for our next sprint. I believe the regex logic is using `\` as an escape character and so in the case of 

```
namespace Tombenevides\SaidQuotes;
```

it is rendering

```
namespace TombenevidesSaidQuotes;
```

See also Tom's description above and the screenshots in the ticket. Thanks!